# Лабораторная работа №3. Классификация текстовых данных

## Цель работы
Получить практические навыки решения задачи классификации текстовых данных в среде Jupiter Notebook. Научиться проводить предварительную обработку текстовых данных, настраивать параметры методов классификации и обучать модели, оценивать точность полученных моделей

## Выполнение работы

## Импорт необходимых библиотек и модулей

In [22]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import GaussianNB
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer 
import numpy as np
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
import pandas as pd
from sklearn.metrics import classification_report,confusion_matrix 
from sklearn.tree import DecisionTreeClassifier
import gensim.downloader
from sklearn.preprocessing import MinMaxScaler

## 1. Загрузить выборки по варианту из лабораторной работы №2. Стемминг проводить не нужно.

In [2]:
categories = ['misc.forsale','sci.med','talk.religion.misc'] 
remove = ('headers', 'footers', 'quotes')
twenty_train = fetch_20newsgroups(subset='train', shuffle=True, random_state=23, categories = categories, remove = remove )
twenty_test = fetch_20newsgroups(subset='test', shuffle=True, random_state=23, categories = categories, remove = remove )

## 2. Используя GridSearchCV произвести предварительную обработку данных и настройку методов классификации в соответствие с заданием, вывести оптимальные значения параметров и результаты классификации модели (полнота, точность, f1-мера и аккуратности) с данными параметрами.

### Наивный байесовский метод

In [4]:
text_clf = Pipeline([('vect', CountVectorizer(max_features= 1000, stop_words = 'english')),
                    ('tfidf', TfidfTransformer(use_idf = True)),
                    ('clf', MultinomialNB (alpha = 0.1)),]) 
parameters = {'vect__max_features': (100,500,1000,5000,10000),
              'vect__stop_words': ('english', None),
              'tfidf__use_idf': (True, False),
             'clf__alpha': (0.1, 1, 2)} 
gs_clf = GridSearchCV(text_clf, parameters, n_jobs = None, cv=3)
gs_clf.fit(twenty_train.data, twenty_train.target)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('vect',
                                        CountVectorizer(max_features=1000,
                                                        stop_words='english')),
                                       ('tfidf', TfidfTransformer()),
                                       ('clf', MultinomialNB(alpha=0.1))]),
             param_grid={'clf__alpha': (0.1, 1, 2),
                         'tfidf__use_idf': (True, False),
                         'vect__max_features': (100, 500, 1000, 5000, 10000),
                         'vect__stop_words': ('english', None)})

In [5]:
print('Выбранные параметры:', gs_clf.best_params_, '\n')
print('Лучший результат классификации:', gs_clf.best_score_, '\n')
print('Результаты по всей сетке параметров: \n')
display(pd.DataFrame(data = gs_clf.cv_results_))

Выбранные параметры: {'clf__alpha': 0.1, 'tfidf__use_idf': True, 'vect__max_features': 10000, 'vect__stop_words': 'english'} 

Лучший результат классификации: 0.9318744838976052 

Результаты по всей сетке параметров: 



mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.236945      0.007332         0.083494        0.006892   
1        0.253379      0.006851         0.097323        0.002863   
2        0.239437      0.011331         0.090699        0.006177   
3        0.264900      0.012882         0.102327        0.014826   
4        0.274479      0.006554         0.100107        0.000220   
5        0.284860      0.004087         0.108494        0.006229   
6        0.274896      0.002085         0.104824        0.003530   
7        0.293310      0.005137         0.110525        0.002095   
8        0.283064      0.004710         0.110573        0.005031   
9        0.294915      0.002351         0.113004        0.003179   
10       0.263989      0.005735         0.092344        0.003417   
11       0.289936      0.009969         0.106017        0.005131   
12       0.269660      0.008972         0.094046        0.002899   
13       0.287651      0.015040         0.109983        0.006491   
14       0.272783      0.009979         0.099957        0.007504   
15       0.293716      0.010137         0.110200        0.003582   
16       0.275812      0.006228         0.101338        0.008209   
17       0.286747      0.010664         0.116081        0.007138   
18       0.280356      0.010153         0.108781        0.000828   
19       0.284854      0.002683         0.111071        0.003842   
20       0.266707      0.007055         0.095349        0.004356   
21       0.248259      0.018862         0.102335        0.008393   
22       0.229819      0.003793         0.085787        0.004999   
23       0.243083      0.003102         0.098494        0.006762   
24       0.271885      0.002569         0.104313        0.008293   
25       0.291243      0.012895         0.112113        0.009046   
26       0.275436      0.002549         0.099773        0.000303   
27       0.293277      0.006973         0.114862        0.004848   
28       0.274118      0.005538         0.108353        0.000782   
29       0.301429      0.001782         0.115586        0.001259   
30       0.262332      0.002686         0.094932        0.004424   
31       0.275848      0.002313         0.104819        0.002973   
32       0.261899      0.004001         0.098839        0.007076   
33       0.291932      0.010438         0.109058        0.006553   
34       0.255081      0.008059         0.106077        0.005426   
35       0.286307      0.012780         0.110430        0.004551   
36       0.264469      0.003994         0.103350        0.003811   
37       0.293058      0.016313         0.116391        0.006933   
38       0.271975      0.007404         0.100889        0.001356   
39       0.293320      0.003245         0.109452        0.001695   
40       0.260556      0.003691         0.095166        0.010889   
41       0.264235      0.007984         0.100483        0.006893   
42       0.261126      0.009503         0.091306        0.006754   
43       0.279740      0.003614         0.105643        0.003881   
44       0.269632      0.005119         0.097845        0.003490   
45       0.287052      0.002668         0.105665        0.006840   
46       0.279112      0.003382         0.109150        0.003775   
47       0.300462      0.010531         0.119855        0.006649   
48       0.277036      0.012093         0.116546        0.004392   
49       0.296434      0.007712         0.129815        0.006746   
50       0.261572      0.004022         0.094165        0.002337   
51       0.284164      0.004135         0.103804        0.002305   
52       0.264111      0.007142         0.099528        0.000822   
53       0.287602      0.015130         0.104787        0.008706   
54       0.263996      0.007430         0.102540        0.003240   
55       0.288567      0.012733         0.104065        0.003145   
56       0.260846      0.004234         0.105030        0.004092   
57       0.294927      0.008237         0.114021        0.00

In [6]:
prediction = gs_clf.best_estimator_.predict(twenty_test.data)
print('Accuracy score: ', gs_clf.best_estimator_.score(twenty_test.data, twenty_test.target)) 
print(classification_report(twenty_test.target, prediction))

Accuracy score:  0.9045323047251688
              precision    recall  f1-score   support

           0       0.95      0.95      0.95       390
           1       0.87      0.93      0.90       396
           2       0.89      0.80      0.84       251

    accuracy                           0.90      1037
   macro avg       0.90      0.89      0.90      1037
weighted avg       0.91      0.90      0.90      1037



### Метод деревьев решений

In [7]:
text_clf = Pipeline([('vect', CountVectorizer(max_features=10000, stop_words='english')),
                    ('tfidf', TfidfTransformer(use_idf = True)),
                    ('clf', DecisionTreeClassifier (criterion='gini', max_depth=15)),])

parameters = {'vect__max_features': (100,500,1000,5000,10000),
              'vect__stop_words': ('english', None),
              'tfidf__use_idf': (True, False),              
              'clf__criterion': ('gini', 'entropy'),
              'clf__max_depth': (5, 15, 50, 100)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1, cv=3)

gs_clf.fit(twenty_train.data, twenty_train.target)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('vect',
                                        CountVectorizer(max_features=10000,
                                                        stop_words='english')),
                                       ('tfidf', TfidfTransformer()),
                                       ('clf',
                                        DecisionTreeClassifier(max_depth=15))]),
             n_jobs=-1,
             param_grid={'clf__criterion': ('gini', 'entropy'),
                         'clf__max_depth': (5, 15, 50, 100),
                         'tfidf__use_idf': (True, False),
                         'vect__max_features': (100, 500, 1000, 5000, 10000),
                         'vect__stop_words': ('english', None)})

In [8]:
print('Выбранные параметры:', gs_clf.best_params_, '\n')
print('Лучший результат классификации:', gs_clf.best_score_, '\n')
print('Результаты по всей сетке параметров: \n')
display(pd.DataFrame(data = gs_clf.cv_results_))

Выбранные параметры: {'clf__criterion': 'gini', 'clf__max_depth': 50, 'tfidf__use_idf': False, 'vect__max_features': 5000, 'vect__stop_words': 'english'} 

Лучший результат классификации: 0.7898493043001714 

Результаты по всей сетке параметров: 



mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0         0.233949      0.003960         0.070442        0.004176   
1         0.255352      0.009464         0.078368        0.000895   
2         0.246201      0.007161         0.068813        0.007269   
3         0.269126      0.005873         0.082679        0.002496   
4         0.258731      0.008534         0.075734        0.005664   
..             ...           ...              ...             ...   
155       1.164245      0.050281         0.215678        0.011691   
156       1.143749      0.030984         0.183180        0.010961   
157       1.284569      0.015856         0.138183        0.008805   
158       1.069419      0.102820         0.122736        0.000273   
159       1.055406      0.086001         0.115316        0.000937   

    param_clf__criterion param_clf__max_depth param_tfidf__use_idf  \
0                   gini                    5                 True   
1                   gini                    5                 True   
2                   gini                    5                 True   
3                   gini                    5                 True   
4                   gini                    5                 True   
..                   ...                  ...                  ...   
155              entropy                  100                False   
156              entropy                  100                False   
157              entropy                  100                False   
158              entropy                  100                False   
159              entropy                  100                False   

    param_vect__max_features param_vect__stop_words  \
0                        100                english   
1                        100                   None   
2                        500                english   
3                        500                   None   
4                       1000                english   
..                       ...                    ...   
155                     1000                   None   
156                     5000                english   
157                     5000                   None   
158                    10000                english   
159                    10000                   None   

                                                params  split0_test_score  \
0    {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.684008   
1    {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.595376   
2    {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.685934   
3    {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.685934   
4    {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.685934   
..                                                 ...                ...   
155  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.672447   
156  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.763006   
157  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.682081   
158  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.743738   
159  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.670520   

     split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0             0.651252           0.644788         0.660016        0.017169   
1             0.651252           0.640927         0.629185        0.024276   
2             0.664740           0.635135         0.661937        0.020833   
3             0.670520           0.642857         0.666437        0.017822   
4             0.662813           0.633205         0.660651        0.021581   
..                 ...                ...              ...             ...   
155           0.693642           0.652510         0.672866        0.016795   
156           0.747592           0.772201         0.760933        0.010153   
157           0.709056           0.687259      

In [9]:
prediction = gs_clf.best_estimator_.predict(twenty_test.data)
print('Accuracy score: ', gs_clf.best_estimator_.score(twenty_test.data, twenty_test.target)) 
print(classification_report(twenty_test.target, prediction))

Accuracy score:  0.7550626808100289
              precision    recall  f1-score   support

           0       0.84      0.82      0.83       390
           1       0.67      0.83      0.74       396
           2       0.81      0.54      0.65       251

    accuracy                           0.76      1037
   macro avg       0.77      0.73      0.74      1037
weighted avg       0.77      0.76      0.75      1037



## 3. Перевести выборку к векторному представлению word embedding согласно варианту.

In [ ]:
glove_model = gensim.downloader.load("word2vec-google-news-300")

In [32]:
glove_model['yellow'] + glove_model['bird']

array([ 0.01123047,  0.15136719, -0.26171875,  0.21899414, -0.06335449,
       -0.40185547,  0.14550781, -0.44970703,  0.046875  ,  0.07531738,
        0.2109375 , -0.65722656, -0.078125  , -0.44921875, -0.25897217,
       -0.3203125 , -0.36621094,  0.10302734, -0.42016602, -0.36914062,
       -0.1328125 , -0.05566406,  0.03430176,  0.02197266, -0.3701172 ,
       -0.07226562, -0.09375   ,  0.22607422,  0.5957031 ,  0.15454102,
       -0.22241211, -0.00683594, -0.04492188, -0.5029297 ,  0.15093994,
        0.04724121,  0.07897949,  0.05029297,  0.00292969,  0.5073242 ,
        0.20507812, -0.21875   ,  0.3544922 , -0.10351562, -0.10375977,
       -0.64941406,  0.14379883, -0.23876953,  0.35595703,  0.4423828 ,
       -0.375     ,  0.01928711,  0.5673828 , -0.06054688,  0.14550781,
        0.22436523,  0.1875    , -0.3876953 ,  0.33190918,  0.1928711 ,
       -0.05273438, -0.2086792 , -0.01953125, -0.22554016, -0.0592041 ,
       -0.1953125 , -0.20141602, -0.12231445, -0.4140625 ,  0.29

In [31]:
glove_model['parrot']

array([-0.00177765, -0.05957031, -0.20507812,  0.16210938, -0.04711914,
       -0.11230469,  0.0279541 ,  0.08349609,  0.11865234,  0.07324219,
       -0.01239014, -0.31835938, -0.27929688, -0.04223633, -0.24902344,
        0.06079102, -0.09326172,  0.13183594, -0.24707031, -0.11865234,
        0.19140625, -0.01977539,  0.04638672, -0.07226562, -0.21875   ,
        0.07568359, -0.06689453,  0.12158203,  0.46484375, -0.10791016,
        0.08056641, -0.08496094, -0.24414062,  0.10058594, -0.08642578,
        0.03088379,  0.1640625 ,  0.06982422, -0.12158203,  0.12597656,
       -0.02416992, -0.19628906,  0.08398438,  0.12597656,  0.06396484,
       -0.38476562, -0.234375  , -0.26367188,  0.18847656,  0.07666016,
       -0.31640625,  0.09570312,  0.546875  , -0.19726562,  0.13378906,
        0.109375  ,  0.05200195, -0.4140625 ,  0.47070312,  0.20019531,
        0.04150391,  0.12597656, -0.15429688, -0.25390625,  0.12792969,
       -0.03613281, -0.14355469, -0.18261719, -0.19238281, -0.19

In [34]:
np.sqrt(sum(pow(a-b, 2) for a, b in zip(glove_model['yellow'] + glove_model['bird'] , glove_model['parrot'])))

4.080615612207368

In [35]:
np.sqrt(sum(pow(a-b, 2) for a, b in zip(glove_model['bird'] , glove_model['parrot'])))

2.967720968714538

In [11]:
print(gensim.downloader.info('word2vec-google-news-300'))

{'num_records': 3000000, 'file_size': 1743563840, 'base_dataset': 'Google News (about 100 billion words)', 'reader_code': 'https://github.com/RaRe-Technologies/gensim-data/releases/download/word2vec-google-news-300/__init__.py', 'license': 'not found', 'parameters': {'dimension': 300}, 'description': "Pre-trained vectors trained on a part of the Google News dataset (about 100 billion words). The model contains 300-dimensional vectors for 3 million words and phrases. The phrases were obtained using a simple data-driven approach described in 'Distributed Representations of Words and Phrases and their Compositionality' (https://code.google.com/archive/p/word2vec/).", 'read_more': ['https://code.google.com/archive/p/word2vec/', 'https://arxiv.org/abs/1301.3781', 'https://arxiv.org/abs/1310.4546', 'https://www.microsoft.com/en-us/research/publication/linguistic-regularities-in-continuous-space-word-representations/?from=http%3A%2F%2Fresearch.microsoft.com%2Fpubs%2F189726%2Frvecs.pdf'], 'che

In [12]:
vectorizer = CountVectorizer(stop_words='english')
train_data = vectorizer.fit_transform(twenty_train['data'])

In [13]:
# Функция для векторизации документов на основе загруженной модели

def text2vec(text_data):
    
    # Векторизуем с помощью обученного CountVectorizer
    X = vectorizer.transform(text_data)
    CV_text_data=pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())
    CV_text_data
    # Создадим датафрейм, в который будем сохранять вектор документа
    glove_data=pd.DataFrame()
    # Словарь из слов в документах
    words_vocab=CV_text_data.columns

    # Пробегаем по каждой строке (по каждому документу)
    for i in range(CV_text_data.shape[0]):

        # Вектор одного документа с размерностью glove-модели:
        one_doc = np.zeros(300) 

        # Пробегаемся по каждому документу, смотрим, какие слова документа присутствуют в нашем словаре
        # Суммируем glove-вектора каждого известного слова в one_doc
        for word in words_vocab[CV_text_data.iloc[i,:] >= 1]:
            if word in glove_model.key_to_index.keys(): 
                #print(word, ': ', glove_model[word])
                one_doc += glove_model[word]
        #print(text_data[i], ': ', one_doc)
        glove_data = pd.concat([glove_data, pd.DataFrame([one_doc])], axis = 0)
    #print('glove_data: ', glove_data)
    return glove_data

In [14]:
train_data_glove = text2vec(twenty_train['data'])
test_data_glove = text2vec(twenty_test['data'])
print('Train data:')
display(train_data_glove.head())
print('Test data:')
display(test_data_glove.head())

Train data:


0         1         2          3          4         5          6    \
0  14.558331  9.545416  2.733467  11.016216 -12.314094  9.158953  15.445671   
0   0.137749  1.712402 -0.197479   1.530396  -0.014902  0.265686  -1.435791   
0   2.706085  2.537697  0.850342   3.615967  -3.675995 -0.456451   4.234619   
0   4.926292  1.444272  0.213455   9.576286  -0.796806  2.127808   4.052650   
0   1.475952  0.684814  0.221619   1.512451  -0.591858 -0.534058   0.440796   

         7          8          9    ...       290       291        292  \
0 -18.185631  13.374420  18.192490  ... -9.447099  1.871529 -15.406441   
0  -1.376923   2.027283   0.910034  ...  0.271820  0.198425  -0.591309   
0  -2.554932   6.118164   1.383286  ... -2.533953  1.241486  -2.920166   
0  -7.413940   5.583755   8.226730  ... -1.806427  3.994720  -8.933167   
0  -1.137451   0.430943   0.068726  ... -0.389160  1.073654  -1.178955   

        293        294       295       296        297       298       299  
0  0.122070 -10.564957 -2.761074 -0.157583 -17.212467  3.946602  5.467724  
0  0.786240   1.242432  0.074219 -1.487610  -1.017090  0.122688  0.184265  
0  0.994637  -1.099411 -4.010574  0.326965  -2.579712  0.622253 -0.259918  
0  1.077606  -3.954956 -0.446594 -1.744965  -6.775208 -1.502594  3.669739  
0 -0.012695   0.056915 -0.185791  1.119141   0.110947  0.024719  0.143188  

[5 rows x 300 columns]

Test data:


0         1         2          3         4         5         6    \
0  0.854553  0.976196 -0.722260   2.083862 -1.827788 -0.299255  0.978882   
0  1.684814 -1.059032 -0.763872   6.322968 -2.066849 -1.066833  1.959778   
0  7.388634  7.640450  5.027078  10.007431 -2.753432  3.941029  9.470535   
0  0.950775  0.222290  0.360779   0.099670 -0.753290  0.474373  0.757568   
0  0.170654  0.253830  1.264982   1.117874 -1.919708  0.264420  2.807373   

        7         8          9    ...       290       291        292  \
0 -1.058044  1.016113   1.735085  ... -0.357933  0.288147  -1.266602   
0 -3.683380  5.411682   2.837074  ... -2.576935  0.981430  -3.440716   
0 -9.354985  7.540771  15.581779  ... -6.331673  1.624714 -13.319984   
0 -1.306641  0.765137   0.287109  ... -0.269653 -0.251221  -0.771225   
0 -1.433578  3.189831   1.972412  ... -1.737061 -0.243774  -0.594482   

        293       294       295       296        297       298       299  
0 -0.298523  0.358078 -0.282227  0.679901  -0.441254  0.436226 -0.978699  
0 -0.711662  2.702026  2.848633 -2.818439  -2.609360  2.692444 -2.559536  
0 -0.528517 -6.369415 -4.737461  1.019753 -12.353575  0.330963  4.278614  
0  0.490601  0.006622  0.493652 -0.095581  -0.651367 -0.260742  1.042969  
0 -0.481964  0.227726 -1.587688  0.215050  -2.338043  1.757324  0.669102  

[5 rows x 300 columns]

In [21]:
X_train

array([[0.66419296, 0.22432005, 0.43836546, ..., 0.66772207, 0.36451412,
        0.26857011],
       [0.48004852, 0.08461568, 0.36235589, ..., 0.91081686, 0.32117523,
        0.16562145],
       [0.51284504, 0.09933508, 0.38952951, ..., 0.88736169, 0.32683712,
        0.1569665 ],
       ...,
       [0.47660176, 0.08207003, 0.40102102, ..., 0.92606153, 0.34494215,
        0.16968595],
       [0.503906  , 0.09087367, 0.3408288 , ..., 0.88356312, 0.33634939,
        0.1904785 ],
       [0.47322037, 0.0766908 , 0.35023045, ..., 0.90008421, 0.30449633,
        0.13760837]])

## 4. Провести обучение и настройку тех же алгоритмов классификации и с теми же параметрами, что и в п.2, но на векторизованной выборке

### Наивный байесовский метод

In [26]:
text_clf = Pipeline([('tfidf', TfidfTransformer(use_idf = True)),
                    ('clf', MultinomialNB (alpha = 0.1)),])  
scaler = MinMaxScaler()
X_train = scaler.fit_transform(train_data_glove)
parameters = {'tfidf__use_idf': (True, False), 'clf__alpha': (0.1, 1, 2)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=None, cv=3, error_score='raise')

gs_clf.fit(X_train, twenty_train.target)

GridSearchCV(cv=3, error_score='raise',
             estimator=Pipeline(steps=[('tfidf', TfidfTransformer()),
                                       ('clf', MultinomialNB(alpha=0.1))]),
             param_grid={'clf__alpha': (0.1, 1, 2),
                         'tfidf__use_idf': (True, False)})

In [27]:
print('Выбранные параметры:', gs_clf.best_params_, '\n')
print('Лучший результат классификации:', gs_clf.best_score_, '\n')
print('Результаты по всей сетке параметров: \n')
display(pd.DataFrame(data = gs_clf.cv_results_))

Выбранные параметры: {'clf__alpha': 0.1, 'tfidf__use_idf': True} 

Лучший результат классификации: 0.5141309765587222 

Результаты по всей сетке параметров: 



mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.060775      0.003408         0.016785        0.000286   
1       0.048801      0.005508         0.014540        0.004380   
2       0.060272      0.003783         0.016831        0.000030   
3       0.049899      0.000431         0.011082        0.004097   
4       0.053433      0.003464         0.013797        0.003732   
5       0.048217      0.002568         0.015540        0.002073   

  param_clf__alpha param_tfidf__use_idf  \
0              0.1                 True   
1              0.1                False   
2                1                 True   
3                1                False   
4                2                 True   
5                2                False   

                                         params  split0_test_score  \
0   {'clf__alpha': 0.1, 'tfidf__use_idf': True}           0.495183   
1  {'clf__alpha': 0.1, 'tfidf__use_idf': False}           0.495183   
2     {'clf__alpha': 1, 'tfidf__use_idf': True}           0.472062   
3    {'clf__alpha': 1, 'tfidf__use_idf': False}           0.472062   
4     {'clf__alpha': 2, 'tfidf__use_idf': True}           0.441233   
5    {'clf__alpha': 2, 'tfidf__use_idf': False}           0.441233   

   split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0           0.545279           0.501931         0.514131        0.022197   
1           0.545279           0.501931         0.514131        0.022197   
2           0.516378           0.467181         0.485207        0.022131   
3           0.516378           0.467181         0.485207        0.022131   
4           0.481696           0.424710         0.449213        0.023939   
5           0.481696           0.424710         0.449213        0.023939   

   rank_test_score  
0                1  
1                1  
2                3  
3                3  
4                5  
5                5

In [28]:
prediction = gs_clf.best_estimator_.predict(test_data_glove)
print('Accuracy:', gs_clf.best_estimator_.score(test_data_glove , twenty_test['target']))
print(classification_report(twenty_test['target'] , prediction))

Accuracy: 0.43297974927675986
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       390
           1       0.40      1.00      0.57       396
           2       0.98      0.22      0.35       251

    accuracy                           0.43      1037
   macro avg       0.46      0.40      0.31      1037
weighted avg       0.39      0.43      0.30      1037



C:\Users\yaasd\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yaasd\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\yaasd\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [29]:
confusion_matrix(twenty_test['target'] , prediction)

array([[  0, 390,   0],
       [  0, 395,   1],
       [  0, 197,  54]], dtype=int64)

### Метод деревьев решений

In [18]:
text_clf = Pipeline([('tfidf', TfidfTransformer(use_idf = True)),
                    ('clf', DecisionTreeClassifier (criterion='gini', max_depth=15)),])

parameters = {'tfidf__use_idf': (True, False),              
              'clf__criterion': ('gini', 'entropy'),
              'clf__max_depth': (5, 15, 50, 100)}

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1, cv=3)

gs_clf.fit(train_data_glove, twenty_train.target)

GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('tfidf', TfidfTransformer()),
                                       ('clf',
                                        DecisionTreeClassifier(max_depth=15))]),
             n_jobs=-1,
             param_grid={'clf__criterion': ('gini', 'entropy'),
                         'clf__max_depth': (5, 15, 50, 100),
                         'tfidf__use_idf': (True, False)})

In [19]:
print('Выбранные параметры:', gs_clf.best_params_, '\n')
print('Лучший результат классификации:', gs_clf.best_score_, '\n')
print('Результаты по всей сетке параметров: \n')
display(pd.DataFrame(data = gs_clf.cv_results_))

Выбранные параметры: {'clf__criterion': 'gini', 'clf__max_depth': 5, 'tfidf__use_idf': True} 

Лучший результат классификации: 0.7866541190240612 

Результаты по всей сетке параметров: 



mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.664189      0.112601         0.060078        0.007237   
1        0.639930      0.159985         0.046085        0.019045   
2        1.200392      0.259588         0.075075        0.022099   
3        0.974941      0.135818         0.042956        0.013880   
4        0.953016      0.138745         0.058772        0.014184   
5        1.005120      0.055353         0.038267        0.012110   
6        1.158371      0.100922         0.074046        0.000423   
7        1.210535      0.054975         0.049531        0.017869   
8        1.369551      0.034717         0.059975        0.017548   
9        1.403559      0.067751         0.062642        0.005652   
10       1.764209      0.067531         0.065323        0.000549   
11       1.715211      0.024845         0.055776        0.005471   
12       1.620110      0.083027         0.046316        0.009852   
13       1.443779      0.066941         0.026646        0.003420   
14       1.322252      0.090000         0.032583        0.001061   
15       1.060531      0.138544         0.021761        0.004498   

   param_clf__criterion param_clf__max_depth param_tfidf__use_idf  \
0                  gini                    5                 True   
1                  gini                    5                False   
2                  gini                   15                 True   
3                  gini                   15                False   
4                  gini                   50                 True   
5                  gini                   50                False   
6                  gini                  100                 True   
7                  gini                  100                False   
8               entropy                    5                 True   
9               entropy                    5                False   
10              entropy                   15                 True   
11              entropy                   15                False   
12              entropy                   50                 True   
13              entropy                   50                False   
14              entropy                  100                 True   
15              entropy                  100                False   

                                               params  split0_test_score  \
0   {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.772640   
1   {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.776493   
2   {'clf__criterion': 'gini', 'clf__max_depth': 1...           0.766859   
3   {'clf__criterion': 'gini', 'clf__max_depth': 1...           0.751445   
4   {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.759152   
5   {'clf__criterion': 'gini', 'clf__max_depth': 5...           0.745665   
6   {'clf__criterion': 'gini', 'clf__max_depth': 1...           0.749518   
7   {'clf__criterion': 'gini', 'clf__max_depth': 1...           0.763006   
8   {'clf__criterion': 'entropy', 'clf__max_depth'...           0.766859   
9   {'clf__criterion': 'entropy', 'clf__max_depth'...           0.768786   
10  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.786127   
11  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.795761   
12  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.772640   
13  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.768786   
14  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.782274   
15  {'clf__criterion': 'entropy', 'clf__max_depth'...           0.784200   

    split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0            0.766859           0.820463         0.786654        0.024023   
1            0.761079           0.814672         0.784081        0.022527   
2            0.763006           0.787645         0.772503        0.010822   
3            0.759152           0.799228         0.769942        0.02094

In [20]:
prediction = gs_clf.best_estimator_.predict(test_data_glove)
print('Accuracy:', gs_clf.best_estimator_.score(test_data_glove, twenty_test.target))
print(classification_report(twenty_test.target, prediction))


Accuracy: 0.7560270009643202
              precision    recall  f1-score   support

           0       0.90      0.79      0.84       390
           1       0.71      0.81      0.76       396
           2       0.64      0.61      0.62       251

    accuracy                           0.76      1037
   macro avg       0.75      0.74      0.74      1037
weighted avg       0.76      0.76      0.76      1037



## 5. Оформить сравнительную таблицу с результатами классификации различными методами с разными настройками. Сделать выводы о наиболее подходящем методе классификации ваших данных с указанием параметров метода и описанием предварительной обработки данных.

| Метод | Векторизированная выборка |                                                               Параметры                                                               | Accuracy | Recall | F1-score |   |   |   |   |   |   |   |   |
|:-----:|:-------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------:|:--------:|:------:|:--------:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|  MNB  |            Нет            |                          'clf__alpha': 0.1, 'tfidf__use_idf': True, 'vect__max_features': 10000,  'vect__stop_words': 'english'                          |   0.9   |  0.89  |   0.9   |   |   |   |   |   |   |   |   |
|       |             Да            |                                                   'clf__alpha': 0.1,      'tfidf__use_idf': True                                                        |   0.46   |  0.4  |   0.31   |   |   |   |   |   |   |   |   |
|   DT  |            Нет            | 'clf__criterion': 'gini', 'clf__max_depth': 50,  'tfidf__use_idf': False, 'vect__max_features': 5000,  'vect__stop_words': 'english' |   0.77   |  0.73  |   0.74   |   |   |   |   |   |   |   |   |
|       |             Да            |                                 'clf__criterion': 'gini', 'clf__max_depth': 5,  'tfidf__use_idf': True                                |   0.75   |  0.74  |   0.74   |   |   |   |   |   |   |   |   |
